src/Workspaces/Core/Portable/FindSymbols/Declarations/DeclarationFinder_AllDeclarations.cs

```diff
@@ -61,48 +62,75 @@ internal static partial class DeclarationFinder
             using var _1 = ArrayBuilder<ISymbol>.GetInstance(out var buffer);
             using var _2 = ArrayBuilder<ISymbol>.GetInstance(out var result);
 
+            // Lazily produce the compilation.  We don't want to incur any costs (especially source generators) if there
+            // are no results for this query in this project.
+            var lazyCompilation = new AsyncLazy<Compilation>(project.GetRequiredCompilationAsync, cacheResult: true);
+
             if (project.SupportsCompilation)
             {
-                var compilation = await project.GetRequiredCompilationAsync(cancellationToken).ConfigureAwait(false);
-
                 // get declarations from the compilation's assembly
-                await AddCompilationDeclarationsWithNormalQueryAsync(
+                await AddCompilationSourceDeclarationsWithNormalQueryAsync(
                     project, query, criteria, buffer, cancellationToken).ConfigureAwait(false);
 
-                // get declarations from directly referenced projects and metadata
-                foreach (var assembly in compilation.GetReferencedAssemblySymbols())
+                // get declarations from directly referenced projects
+                foreach (var projectReference in project.ProjectReferences)
```

Comment:
instead of walking GetReferencedAssemblySymbols (which needs both a compilation, and creates assembly symbols for all references), we can instead just manually walk project-references and metadtareferences, search each (using indices) And only realize the compilation if we find a match.